filters/auth/oauth.go

```diff
@@ -0,0 +1,429 @@
+package auth
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/filters"
+	logfilter "github.com/zalando/skipper/filters/log"
+	"golang.org/x/oauth2"
+)
+
+type roleCheckType int
+
+const (
+	checkAnyScopes roleCheckType = iota
+	checkAllScopes
+	checkAnyKV
+	checkAllKV
+	checkUnknown
+)
+
+type rejectReason string
+
+const (
+	missingBearerToken rejectReason = "missing-bearer-token"
+	authServiceAccess  rejectReason = "auth-service-access"
+	invalidToken       rejectReason = "invalid-token"
+	invalidRealm       rejectReason = "invalid-realm"
+	invalidScope       rejectReason = "invalid-scope"
+)
+
+// TODO: discuss these names, because these are the filter names used by the endusers
+const (
+	AuthAnyScopeName = "authAnyScope"
+	AuthAllScopeName = "authAllScope"
+	AuthAnyKVName    = "authAnyKV"
+	AuthAllKVName    = "authAllKV"
+	AuthUnknown      = "authUnknown"
```

Comment:
I think `oauth2` should be in the names. `auth` is a bit too generic.

Reply:
will be done in the next push